lib/nmatrix/shortcuts.rb

```diff
@@ -275,6 +275,61 @@ def diagonal(entries, opts={})
     alias :diag :diagonal
     alias :diagonals :diagonal
 
+    # Generate a block-diagonal NMatrix from the supplied 2D square matrices.
+    #
+    # * *Arguments*
+    #   - +*params+ -> An array that collects all arguments passed to the method. The method
+    #                  can receive any number of arguments. Optionally, the last entry of +params+ is 
+    #                  a hash of options from NMatrix#initialize. All other entries of +params+ are 
+    #                  the blocks of the desired block-diagonal matrix, which are supplied
+    #                  as square 2D NMatrix objects.
+    # * *Returns*
+    #   - NMatrix of block-diagonal form filled with specified matrices
+    #     as the blocks along the diagonal.
+    #
+    # * *Example* 
+    #
+    #  a = NMatrix.new([2,2],[1,2,3,4])
+    #  b = NMatrix.new([1,1],[123],dtype: :int32)
+    #  c = NMatrix.new([3,3],[1,2,3,1,2,3,1,2,3], dtype: :float64)
+    #  m = NMatrix.block_diagonal(a,b,c,dtype: :int64, stype: :yale)
+    #      => 
+    #      [
+    #        [1, 2,   0, 0, 0, 0]
+    #        [3, 4,   0, 0, 0, 0]
+    #        [0, 0, 123, 0, 0, 0]
+    #        [0, 0,   0, 1, 2, 3]
+    #        [0, 0,   0, 1, 2, 3]
+    #        [0, 0,   0, 1, 2, 3]
+    #      ]
+    #
+    def block_diagonal(*params)
+      options = params.last.is_a?(Hash) ? params.pop : {}
+
+      block_sizes = [] #holds the size of each matrix block
+      params.each do |b|
+        raise ArgumentError, "Only NMatrix objects allowed" unless b.is_a?(NMatrix)
```

Comment:
Well, in this specific case you can:

```ruby
a = NMatrix.ones([2, 2])
a100 = Array.new(100) { a }
m = NMatrix.block_diagonal(*a100) # The array elements will be sent to the function as parameters.
```

Thus not needing a change in your function. :)